<a href="https://colab.research.google.com/github/Amirgh8080/SDG_Test/blob/main/SDG.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Testing the implmentation of the SDG model

# 1.Preparing the VM

## Installing the nessassaries libraries suggested by the paper's authors

In [1]:
!pip install numpy==1.14.6 scipy==1.2.1 pytorch==1.6.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 55.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.1/23.1 MB 88.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
ERROR: Ignored the following versions that require a different python version: 1.6.2 Requires-Python >=3.7,<3.10; 1.6.3 Requires-Python >=3.7,<3.10; 1.7.0 Requires-Python >=3.7,<3.10; 1.7.1 Requires-Python >=3.7,<3.10
ERROR: Could not find a version that satisfies the requirement pytorch==1.6.0 (from versions: 0.1.2, 1.0.2)
ERROR: No matching distribution found for pytorch==1.6.0


## Extracting the model files

In [2]:
!unzip SDG-main.zip

Archive:  SDG-main.zip
7a09ae77466aca8ecb68300104d6880fab295412
   creating: SDG-main/
  inflating: SDG-main/README.md      
   creating: SDG-main/code/
  inflating: SDG-main/code/README.md  
  inflating: SDG-main/code/main.py   
   creating: SDG-main/code/sdg/
   creating: SDG-main/code/sdg/data/
  inflating: SDG-main/code/sdg/data/citeseer.npz  
  inflating: SDG-main/code/sdg/data/cora_ml.npz  
  inflating: SDG-main/code/sdg/data/io.py  
  inflating: SDG-main/code/sdg/data/pubmed.npz  
  inflating: SDG-main/code/sdg/data/sparsegraph.py  
   creating: SDG-main/code/sdg/pytorch_code/
  inflating: SDG-main/code/sdg/pytorch_code/agnostic_model.py  
  inflating: SDG-main/code/sdg/pytorch_code/earlystopping.py  
  inflating: SDG-main/code/sdg/pytorch_code/preprocessing.py  
  inflating: SDG-main/code/sdg/pytorch_code/propagation.py  
  inflating: SDG-main/code/sdg/pytorch_code/training.py  
  inflating: SDG-main/code/sdg/pytorch_code/utils.py  
   creating: SDG-main/paper/
  inflating: SDG

# 2.Testing Model

## Testing the model on Cora_ML Dataset

In [22]:
%%time
!python SDG-main/code/main.py

2024-11-27 21:43:01: agnostic_model: {'hiddenunits': [64], 'drop_prob': 0.5, 'propagation': PPRExact()}
2024-11-27 21:43:01: PyTorch seed: 1837425462
/content/SDG-main/code/sdg/pytorch_code/utils.py:74: UserWarning: torch.sparse.SparseTensor(indices, values, shape, *, device=) is deprecated.  Please use torch.sparse_coo_tensor(indices, values, shape, dtype=, device=). (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:651.)
  return torch.sparse.FloatTensor(
2024-11-27 21:43:03: Epoch 0: Train loss = 2.00, train acc = 12.9, early stopping loss = 1.96, early stopping acc = 37.8 (0.361 sec)
2024-11-27 21:43:03: Epoch 20: Train loss = 1.94, train acc = 71.4, early stopping loss = 1.95, early stopping acc = 55.0 (0.440 sec)
2024-11-27 21:43:03: Epoch 40: Train loss = 1.90, train acc = 79.3, early stopping loss = 1.94, early stopping acc = 58.6 (0.265 sec)
2024-11-27 21:43:04: Epoch 60: Train loss = 1.84, train acc = 83.6, early stopping loss = 1.92, early stopping acc = 65.2 (0.40

## Testing the model on Citeseer Dataset

In [23]:
%%time
!python SDG-main/code/main.py

/content/SDG-main/code/sdg/data/sparsegraph.py:397: UserWarning: 124 self loops removed
  warnings.warn("{0} self loops removed".format(num_self_loops))
2024-11-27 21:44:54: agnostic_model: {'hiddenunits': [64], 'drop_prob': 0.5, 'propagation': PPRExact()}
2024-11-27 21:44:54: PyTorch seed: 1700194395
/content/SDG-main/code/sdg/pytorch_code/utils.py:74: UserWarning: torch.sparse.SparseTensor(indices, values, shape, *, device=) is deprecated.  Please use torch.sparse_coo_tensor(indices, values, shape, dtype=, device=). (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:651.)
  return torch.sparse.FloatTensor(
2024-11-27 21:44:56: Epoch 0: Train loss = 1.85, train acc = 17.5, early stopping loss = 1.81, early stopping acc = 32.6 (0.380 sec)
2024-11-27 21:44:56: Epoch 20: Train loss = 1.78, train acc = 53.3, early stopping loss = 1.78, early stopping acc = 55.0 (0.388 sec)
2024-11-27 21:44:56: Epoch 40: Train loss = 1.73, train acc = 64.2, early stopping loss = 1.74, early stoppi

## Testing the model on Pubmed Dataset

In [24]:
%%time
!python SDG-main/code/main.py

2024-11-27 21:55:07: agnostic_model: {'hiddenunits': [64], 'drop_prob': 0.5, 'propagation': PPRExact()}
2024-11-27 21:55:07: PyTorch seed: 554409738
/content/SDG-main/code/sdg/pytorch_code/utils.py:74: UserWarning: torch.sparse.SparseTensor(indices, values, shape, *, device=) is deprecated.  Please use torch.sparse_coo_tensor(indices, values, shape, dtype=, device=). (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:651.)
  return torch.sparse.FloatTensor(
2024-11-27 21:55:09: Epoch 0: Train loss = 1.15, train acc = 33.3, early stopping loss = 1.12, early stopping acc = 65.6 (0.421 sec)
2024-11-27 21:55:25: Epoch 20: Train loss = 1.07, train acc = 83.3, early stopping loss = 1.08, early stopping acc = 79.4 (15.290 sec)
2024-11-27 21:55:34: Epoch 40: Train loss = 0.98, train acc = 81.7, early stopping loss = 1.02, early stopping acc = 80.4 (9.456 sec)
2024-11-27 21:55:36: Epoch 60: Train loss = 0.88, train acc = 90.0, early stopping loss = 0.96, early stopping acc = 80.6 (1.87

# 3.Conclusion

model was tested on various datasets, yielding the following results:

##Cora_ML Dataset:
For the agnostic model, training commenced at 21:43:01, with the initial training accuracy at 12.9% and a loss of 2.00. Over the course of 2261 epochs, the model exhibited a remarkable improvement, achieving a training accuracy of 100% by epoch 640 and maintaining high performance with an early stopping accuracy of 84.4% and a validation accuracy of 84.4%. The total training time for the PPNP was approximately 29.97 seconds, with the entire training process taking around 40.8 seconds.

Subsequently, the fine-tuned model was initialized at 21:43:32, where it quickly reached a training accuracy of 100% by epoch 20. The model ultimately achieved an early stopping accuracy of 85.4% and a validation accuracy of 84.8%, with the training process for the SDG model costing approximately 7.68 seconds

##Citeseer Dataset:
The training for the agnostic model began at 21:44:54, with an initial training accuracy of 17.5% and a loss of 1.85. Over the course of 1003 epochs, the model demonstrated substantial improvement, achieving a peak training accuracy of 100% by epoch 940. The early stopping accuracy settled at 73.8%, while the validation accuracy reached 71.4%. The total training time for the PPNP was approximately 14.84 seconds, with the entire process taking around 25.6 seconds.

Following this, the fine-tuned model was initialized at 21:45:10, where it quickly reached a training accuracy of 94.2% in the first epoch. Throughout its training, it maintained competitive performance, achieving an early stopping accuracy of 73.2% and a validation accuracy of 70.0%. The combined time for generating the new graph and training the SDG model was approximately 8.36 seconds.

##Pubmed Dataset:
The training for the agnostic model commenced at 21:55:07, starting with a training accuracy of 33.3% and a loss of 1.15. Over the course of 1282 epochs, the model showed impressive improvement, achieving a peak training accuracy of 100% by epoch 700. The early stopping accuracy reached 84.2%, with a validation accuracy of 81.0%. The total time taken for training the PPNP was approximately 656.96 seconds, while the entire process, including graph generation, took about 21 minutes and 20 seconds.

Following this, the fine-tuned model was initialized at 22:06:45, where it began with a training accuracy of 100% in the first epoch. Throughout its training, it maintained high performance, achieving an early stopping accuracy of 84.0% and a validation accuracy of 81.6%. The total time for generating the new graph and training the SDG model was approximately 620.90 seconds.

#Insights:

The model demonstrated high training accuracy across epochs, achieving stable validation accuracy for Cora_ML, Citerseer and pubmed Datasets.
The only problem is that the time the model took to process the pubmed showed significant difference from what the model reported(the difference in others was milliseconds).